python/numpy.h

```diff
@@ -12,24 +13,89 @@
 #include "pybind11.h"
 
 namespace py = pybind11;
+
 using namespace scipp;
 
-template <class T, class View>
+/// Map C++ types to Python types to perform conversion between scipp containers
+/// and numpy arrays.
+template <class T> struct ElementTypeMap {
+  using CppType = T;
+  using PyType = T;
+  constexpr static bool convert = false;
+
+  static void check_assignable(const py::object &, const units::Unit &) {}
+};
+
+template <> struct ElementTypeMap<scipp::core::time_point> {
+  using CppType = scipp::core::time_point;
+  using PyType = int64_t;
+  constexpr static bool convert = true;
+
+  static void check_assignable(const py::object &obj, units::Unit unit);
+};
+
+template <bool convert, class Source, class Destination>
+void copy_element(const Source &src, Destination &&dst) {
+  if constexpr (convert) {
+    dst = std::remove_reference_t<Destination>{src};
+  } else {
+    std::forward<Destination>(dst) = src;
+  }
+}
+
+/// Cast a py::object referring to an array to py::array_t<auto> if supported.
+/// Otherwise, copies the contents into a std::vector<auto>.
+template <class T>
+auto cast_to_array_like(const py::object &obj, const units::Unit unit) {
+  using TM = ElementTypeMap<T>;
+  using PyType = typename TM::PyType;
+  if constexpr (std::is_same_v<T, core::time_point>) {
+    TM::check_assignable(obj, unit);
```

Comment:
Is this some trick I don't see right now, or is it doing nothing?

Reply:
Which part? The `ElementTypeMap` basically says that `core::time_point` is constructed from `int64_t` in numpy arrays.

`check_assignable` makes sure that the units match.

Reply:
Ah, I was looking at the non-specialized version, and saw the empty function body, which led to the confusion.